csp_billing_adapter/adapter.py

```diff
@@ -128,27 +142,56 @@ def event_loop_handler(
     """Perform the event loop processing actions."""
     log.info('Starting event loop processing')
 
-    usage = hook.get_usage_data(config=config)
-    log.debug('Retrieved usage data: %s', usage)
-
-    add_usage_record(hook, config, usage)
-
-    cache = hook.get_cache(config=config)
-    now = get_now()
-
-    log.debug(
-        "Now: %s, Next Reporting Time: %s, Next Bill Time: %s",
-        date_to_string(now),
-        cache['next_reporting_time'],
-        cache['next_bill_time']
-    )
+    try:
+        usage = hook.get_usage_data(config=config)
+    except Exception as exc:
+        log.warning("Failed to retrieve usage data: %s", str(exc))
+        now = get_now()
+        hook.update_csp_config(
+            config=config,
+            csp_config={
+                'timestamp': date_to_string(now),
+                'billing_api_access_ok': False,
+                'errors': [
+                    f'Usage data retrieval failed: {exc}'
+                ]
+            },
+            replace=False
+        )
+        # NOTE:
+        # If this happens during initial deployment timeframe, failing
+        # to retrieve usage data here could simple mean that there is
+        # nothing to bill for yet so not proceeding to metering makes
+        # sense.
+        # If this occurs for an existing deployment, with valid usage
+        # records waiting to be billed, continuing on to metering is
+        # worth considering.
+        # However, doing so will currently result in the error state
+        # that was saved in the csp_config being lost, as it would be
+        # overwritten with a success state, or a different error.
```

Comment:
A number of calls inside else clause could also fall over. It might make for cleaner exception handling if any early exit of event loop is handled with a "continue". Also the rest of the loop is not necessarily tied to one function. In this case if the adapter cannot get usage data it probably shouldn't prevent the adapter from billing if it's time to bill and there's something in cache to bill. To avoid loss of state it could be tracked and only updated once at the end of loop.